File: ruby/rc.rb
# frozen_string_literal: true

require_relative 'base'

class RcParser < ParserBase
    PARSER_PROC = proc {
        single_quoted = proc {
            between proc { take_exact %[\'] },
                    proc { take_exact %[\'] }, proc {
                %[\'] + (take_while proc { |c| c != %[\'] }) + %[\']
            }
        }

        double_quoted = proc {
            between proc { take_exact %[\"] },
                    proc { take_exact %[\"] }, proc {
                escaped = false
                %[\"] + (take_while proc { |c|
                    (c == %[\\] || (c == %[\"] && escaped)) \
                        ? (escaped = !escaped; true)
                        : (c != %[\"])
                }) + %[\"]
            }
        }

        unquoted = proc {
            peek[/[^\#[:space:]]/] \
                ? (take_while proc { |c| c[/[^\#[:space:]]/] })
                : quit
        }

        attribute = proc { take_while proc { |c| c[/[[:alnum:]_]/] } }

        blank = proc { take_while proc { |c| c[/[[:blank:]]/] } }

        eol = proc { take_exact %[\n] }

        blank_eol = proc { blank.call + eol.call }

        comment = proc {
            Comment[ (blank.call + (take_exact %[\#]) \
                                 + (take_while proc { |c| c != %[\n] }) \
                                 + eol.call) ]
        }

        pair = proc {
            Pair[ (one_of proc { blank.call + (take_exact %[export]) + blank.call },
                          blank),
                  attribute.call,
                  (take_exact %[\=]),
                  (one_of single_quoted,
                          double_quoted,
                          unquoted,
                          blank),
                  (one_of comment,
                          blank_eol,
                          blank) ]
        }

        Sequence[ *(zero_or_more proc { one_of pair,
                                               comment,
                                               blank_eol }) ]
    }

    def initialize(input)
        super(input, PARSER_PROC)
    end

    def render(node = @parsed)
        node.each_with_object(+'') do |v, acc|
            case
            when v.is_a?(String)
                acc << v
            when v.is_a?(Pair)
                acc << v.join
            else
                raise InvalidTree
            end
        end
    end

    def put(path, value)
        atrb, atrb_i = ypath path, wildcards: false
        s = searchable
        case
        # add a new pair directly at the root level
        when s[atrb].nil? || (!atrb_i.nil? && atrb_i == 0)
            @parsed << Pair[
                %[], atrb, %[\=], value.to_s, %[\n]
            ]

        # require paths to be unequivocal
        when atrb_i.nil? && s[atrb].length > 1
            raise AmbiguousMatch

        # fail when nothing found
        when (s = s.dig(atrb, atrb_i.nil? ? 0 : (atrb_i - 1))).nil?
            raise EmptyMatch

        # the pattern must resolve into a pair
        when !s.is_a?(Pair)
            raise InvalidPath

        # update the value (root level)
        else
            s[3] = value.to_s
        end
    end

    def drop(path, value = nil)
        atrb, atrb_i = ypath path, wildcards: true
        (recurse = proc { |node| # NOTE: never use default values here!
            case
            when node.is_a?(Pair) && (value.nil? || node[3] == value.to_s)
                # delete first found (should be only one)
                @parsed.delete_at(@parsed.index { |x| x.object_id == node.object_id } || @parsed.length)
            when node.is_a?(Sequence)
                node.each { |v| recurse.call v }
            when node.is_a?(Lookup)
                node.each { |_, v| recurse.call v }
            end
        }).call (filtered patterns: [atrb, nil], indices: [atrb_i, nil])
    end

    private

    def ypath(path, wildcards: true)
        wild = wildcards ? %[\*] : %[]
        regx = %r{ ^
                   ( [[:alnum:]_#{wild}]+ )
                   (?:
                       \[ ( [[:digit:]]* ) \]
                   )?
                   $ }x
        raise InvalidPath if (m = regx.match(path)).nil?

        atrb   = m[1]
        atrb_i = m[2].to_s.empty? ? nil : m[2].to_i
        raise InvalidPath if atrb.nil? \
                          || (!atrb_i.nil? && atrb_i.to_i < 0)

        [atrb, atrb_i, nil, nil]
    end

    def searchable(node = @parsed)
        node.each_with_object(Lookup[]) do |v, acc|
            case
            when v.is_a?(Pair)
                (acc[v[1]] ||= Sequence[]) << v
            end
        end.then do |node|
            # at this point it's possible to easily compute and attach index metadata
            (recurse = proc { |node, index| # NOTE: never use default values here!
                case
                when node.is_a?(Pair)
                    node.setm(:@_index_, index)
                when node.is_a?(Sequence)
                    # when the node has less than 2 items the value of @_index_ must be unset
                    node.each_with_index { |v, i| recurse.call v, ((node.length > 1) ? (i + 1) : nil) }
                when node.is_a?(Lookup)
                    node.setm(:@_index_, index)
                    node.each { |_, v| recurse.call v, nil }
                end
            }).call node, nil
            node
        end
    end
end

File: python/rc.py
import re
from base import ParserBase

class RcParser(ParserBase):
    def parser_proc(self):
        def single_quoted():
            return self.between(lambda: self.take_exact("'"),
                                lambda: self.take_exact("'"),
                                lambda: "'" + self.take_while(lambda c: c != "'") + "'")

        def double_quoted():
            def pred(c):
                if c == '\\' or (c == '"' and pred.__dict__['escaped']):
                    pred.__dict__['escaped'] = not pred.__dict__['escaped']
                    return True
                else:
                    return c != '"'
            pred.__dict__['escaped'] = False
            return self.between(lambda: self.take_exact('"'),
                                lambda: self.take_exact('"'),
                                lambda: '"' + self.take_while(pred) + '"')

        def unquoted():
            def pred(c):
                return not (c == '#' or c.isspace())
            return self.take_while(pred) if pred(self.peek()) else self.quit()

        def attribute():
            return self.take_while(lambda c: c == '_' or c.isalnum())

        def blank():
            return self.take_while(lambda c: c in [' ', '\t'])

        def eol():
            return self.take_exact('\n')

        def blank_eol():
            return blank() + eol()

        def comment():
            return self.Comment.from_str(
                blank() + self.take_exact('#')
                        + self.take_while(lambda c: c != '\n')
                        + eol()
            )

        def pair():
            return self.Pair.from_args(
                self.one_of(lambda: blank() + self.take_exact('export') + blank(),
                            blank),
                attribute(),
                self.take_exact('='),
                self.one_of(single_quoted,
                            double_quoted,
                            unquoted,
                            blank),
                self.one_of(comment,
                            blank_eol,
                            blank)
            )

        return self.Sequence.from_args(
            *(self.zero_or_more(lambda: self.one_of(pair,
                                                    comment,
                                                    blank_eol)))
        )

    def __init__(self, inp):
        super().__init__(inp)

    def render(self, node=None):
        acc = ''
        for v in (node or self.parsed):
            if isinstance(v, str):
                acc += v
            elif isinstance(v, self.Pair):
                acc += ''.join(v)
            else:
                raise self.InvalidTree
        return acc

    # due to overwhelming complexity we decided not to use recurrence here.. xD
    def put(self, path, value):
        atrb, atrb_i, _, _ = self._ypath(path, wildcards=False)
        s = self._searchable()

        # add a new pair directly at the root level
        if s.get(atrb) is None or (atrb_i is not None and atrb_i == 0):
            self.parsed.append(self.Pair.from_args(
                '', atrb, '=', str(value), '\n'
            ))

        # require paths to be unequivocal
        elif atrb_i is None and len(s[atrb]) > 1:
            raise self.AmbiguousMatch

        # fail when nothing found
        elif s.get(atrb) is None or (s := s[atrb][0 if atrb_i is None else (atrb_i - 1)]) is None:
            raise self.EmptyMatch

        # the pattern must resolve into a pair
        elif not isinstance(s, self.Pair):
            raise self.InvalidPath

        # update the value (root level)
        else:
            s[3] = str(value)

    def drop(self, path, value=None):
        atrb, atrb_i, _, _ = self._ypath(path, wildcards=True)
        def recurse(node):
            if isinstance(node, self.Pair) and (value is None or node[3] == str(value)):
                for i, x in enumerate(self.parsed):
                    if id(x) == id(node):
                        self.parsed.pop(i)
                        break
            elif isinstance(node, self.Sequence):
                for v in node:
                    recurse(v)
            elif isinstance(node, self.Lookup):
                for v in node.values():
                    recurse(v)
        recurse(self._filtered(patterns=[atrb, None], indices=[atrb_i, None]))

    def _ypath(self, path, wildcards=True):
        wild = '*' if wildcards else ''
        regx = rf'''(?x)
            ^
            ( [A-Za-z0-9_{wild}]+ )
            (?:
                \[ ( [0-9]* ) \]
            )?
            $
        '''
        if (m := re.match(regx, path)) is None:
            raise self.InvalidPath

        atrb   = m[1]
        atrb_i = None if len(m[2] or '') == 0 else int(m[2])
        if atrb is None or (atrb_i is not None and int(atrb_i) < 0):
            raise self.InvalidPath

        return atrb, atrb_i, None, None

    def _searchable(self, node=None):
        def recurse(node, index):
            if isinstance(node, self.Pair):
                node.setm('_index_', index)
            elif isinstance(node, self.Sequence):
                for i, v in enumerate(node):
                    recurse(v, (i + 1) if (len(node) > 1) else None)
            elif isinstance(node, self.Lookup):
                node.setm('_index_', index)
                for v in node.values():
                    recurse(v, None)
        acc = self.Lookup()
        for v in (node or self.parsed):
            if isinstance(v, self.Pair):
                acc[v[1]] = acc.get(v[1], self.Sequence())
                acc[v[1]].append(v)
        recurse(acc, None)
        return acc
